# DATA WRANGLING STEP

## Import libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np

## LOAD DATA

In [2]:
filepath = "./LNGDATA_M.xlsx"
df = pd.read_excel(filepath, skiprows=1)
df.head()

Unnamed: 0  Unnamed: 1      LNG Flow LNG Rundown Temperature  \
0         NaN         NaN   12096.84375             -150.820312   
1         NaN         NaN  12093.041016             -150.834961   
2         NaN         NaN  12115.166016             -150.834961   
3         NaN         NaN  12121.388672             -150.825195   
4         NaN         NaN  12201.246094             -150.849609   

  MR Operating Margin PR Operating Margin MR Compressor Suction Pressure  \
0            2.859555           18.209555                       2.317578   
1            2.921342           19.027306                       2.327148   
2            3.106348           19.838921                       2.329688   
3            2.908482           19.634964                        2.34707   
4             3.01809           19.833036                       2.341211   

  MCHE Warm bypass Temperature MR return Temperature  \
0                   -41.851562             -36.96875   
1                   -41.924805            -37.035156   
2                   -42.402344            -37.070312   
3                   -42.188477            -37.117188   
4                   -42.080078            -37.082031   

  MR Expander inlet Temperature  ... Unnamed: 33  \
0                   -127.460938  ...     No Data   
1                   -127.407227  ...     No Data   
2                   -127.456055  ...     No Data   
3                   -127.275391  ...     No Data   
4                   -127.250977  ...     No Data   

  Low Pressure Propane kettle level  \
0                         57.330078   
1                         56.773438   
2                         57.082031   
3                         57.013672   
4                         56.988281   

  Medium Pressure Propane/Natural Gas kettle level  \
0                                         64.28125   
1                                        64.289062   
2                                        64.048828   
3                                        64.261719   
4                                         64.34375   

  High Pressure Propane/Natural Gas kettle level  \
0                                      64.443359   
1                                      63.486328   
2                                      64.123047   
3                                      64.015625   
4                                      64.001953   

  High High Pressure Propane/Mixed Refrigerant kettle level  \
0                                          65.195312          
1                                          65.564453          
2                                          65.482422          
3                                          65.261719          
4                                          65.244141          

  High Pressure Propane/Mixed Refrigerant kettle level  \
0                                          65.019531     
1                                          65.255859     
2                                          65.126953     
3                                          65.085938     
4                                           65.15625     

  Medium Pressure Propane/Mixed Refrigerant kettle level  \
0                                          64.574219       
1                                          64.496094       
2                                          64.095703       
3                                          64.152344       
4                                          64.169922       

  Low Pressure Propane/Mixed Refrigerant kettle level  \
0                                          56.248047    
1                                          56.039062    
2                                          56.078125    
3                                          56.035156    
4                                          56.042969    

  MCHE Warm bypass temperature Natural Gas Warm Bundle to scrub column  
0                   -41.851562                              -44.722656  
1                   -41.924805            

This data is likely to contain missing values, erratic values and unwanted values.

This section focuses on handling all the missing values, converting correcting the data format

### - Handle missing values, and Errors in data

In [3]:
df.shape

(70129, 43)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70129 non-null  object 
 3   LNG Rundown Temperature                                    70129 non-null  object 
 4   MR Operating Margin                                        70129 non-null  object 
 5   PR Operating Margin                                        70129 non-null  object 
 6   MR Compressor Suction Pressure                             70129 non-null  object 
 7   MCHE Warm bypass Temperature                               70129 non-null  object 
 8   MR ret

---
First, We will replace all the "No Data" values in the data to null values

In [5]:
df.replace("No Data", np.nan, inplace=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70129 non-null  object 
 3   LNG Rundown Temperature                                    70129 non-null  object 
 4   MR Operating Margin                                        70129 non-null  object 
 5   PR Operating Margin                                        70129 non-null  object 
 6   MR Compressor Suction Pressure                             70129 non-null  object 
 7   MCHE Warm bypass Temperature                               70129 non-null  object 
 8   MR ret

---
Let's find out the erratic values in LNG Flow column

In [7]:
error1 = list(map(lambda x: isinstance(x, str), df["LNG Flow"]))
df['LNG Flow'].loc[error1].unique()

array(['I/O Timeout', 'Invalid Data', 'Scan Off', 'Not Connect',
       'Scan Timeout', 'Bad', 'Comm Fail'], dtype=object)

In [8]:
error_values1 = df['LNG Flow'].loc[error1].unique()

In [9]:
df.replace(error_values1, np.nan, inplace= True)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70010 non-null  float64
 3   LNG Rundown Temperature                                    70015 non-null  float64
 4   MR Operating Margin                                        70086 non-null  float64
 5   PR Operating Margin                                        70090 non-null  float64
 6   MR Compressor Suction Pressure                             70039 non-null  float64
 7   MCHE Warm bypass Temperature                               70043 non-null  float64
 8   MR ret

---
Lets find out the error values in "MCHE Cold bundle dp" column

In [12]:
error2 = list(map(lambda x: isinstance(x, str), df["MCHE Cold bundle dp"]))
df['MCHE Cold bundle dp'].loc[error2].unique()

array(['Out of Serv'], dtype=object)

In [13]:
error_values2 = df['MCHE Cold bundle dp'].loc[error2].unique()

In [14]:
df.replace(error_values2, np.nan, inplace= True)

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70010 non-null  float64
 3   LNG Rundown Temperature                                    70015 non-null  float64
 4   MR Operating Margin                                        70086 non-null  float64
 5   PR Operating Margin                                        70090 non-null  float64
 6   MR Compressor Suction Pressure                             70039 non-null  float64
 7   MCHE Warm bypass Temperature                               70043 non-null  float64
 8   MR ret

---
Lets find out the errors in "Volumetric MR Flow to MR comp. discharge" column

In [16]:
error3 = list(map(lambda x: isinstance(x, str), df["Volumetric MR Flow to MR comp. discharge"]))
df['Volumetric MR Flow to MR comp. discharge'].loc[error3].unique()

array(['Unit Down'], dtype=object)

In [17]:
error_values3 = df['Volumetric MR Flow to MR comp. discharge'].loc[error3].unique()

In [18]:
df.replace(error_values3, np.nan, inplace= True)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70010 non-null  float64
 3   LNG Rundown Temperature                                    70015 non-null  float64
 4   MR Operating Margin                                        70086 non-null  float64
 5   PR Operating Margin                                        70090 non-null  float64
 6   MR Compressor Suction Pressure                             70039 non-null  float64
 7   MCHE Warm bypass Temperature                               70043 non-null  float64
 8   MR ret

---
Lets find out the errors in "Percentage Nitrogen in MR " column

In [20]:
error4 = list(map(lambda x: isinstance(x, str), df["Percentage Nitrogen in MR"]))
df['Percentage Nitrogen in MR'].loc[error4].unique()

array(['Bad Input'], dtype=object)

In [21]:
error_values4 = df['Percentage Nitrogen in MR'].loc[error4].unique()

In [22]:
df.replace(error_values4, np.nan, inplace= True)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70129 entries, 0 to 70128
Data columns (total 43 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   Unnamed: 0                                                 0 non-null      float64
 1   Unnamed: 1                                                 0 non-null      float64
 2   LNG Flow                                                   70010 non-null  float64
 3   LNG Rundown Temperature                                    70015 non-null  float64
 4   MR Operating Margin                                        70086 non-null  float64
 5   PR Operating Margin                                        70090 non-null  float64
 6   MR Compressor Suction Pressure                             70039 non-null  float64
 7   MCHE Warm bypass Temperature                               70043 non-null  float64
 8   MR ret

---
Finally, We hve successfuly removed all the errors in the data.
But we have two issues left:
- Removing the unnamed columns
- Eliminating Null values

### - Removing the unnamed columns

In [24]:
df.columns

Index(['Unnamed: 0', 'Unnamed: 1', 'LNG Flow', 'LNG Rundown Temperature',
       'MR Operating Margin', 'PR Operating Margin',
       'MR Compressor Suction Pressure', 'MCHE Warm bypass Temperature',
       'MR return Temperature', 'MR Expander inlet Temperature',
       'LNG Shell Temp Top Bundle', 'LMR JT Valve Valve Opening',
       'MCHE Cold bundle dp', 'Volumetric MR Flow to MR comp. discharge',
       'HMR Flow ', 'HMR/MR Ratio', 'LNG Flow Ex MCHE',
       'MCHE Warm Bypass Temperature', 'MR IGV Position', 'MR Vessel Level',
       'MR Vessel Pressure', 'Percentage Nitrogen in MR',
       'Percentage Methane in MR', 'Percentage Ethane in MR',
       'Percentage Propane in MR', 'Percentage Butane in MR',
       'Nitrogen Make-up flow', 'Methane Make-up flow', 'Ethane Make-up flow',
       'Propane Make-up flow', 'Ambient Condition',
       'Scrub Column inlet Temperature',
       'Natural Gas temp at MCHE warm bundle', 'Unnamed: 33',
       'Low Pressure Propane kettle level',
  

In [25]:
columns_to_drop = ['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 33']
df.drop(columns_to_drop, axis = 1, inplace=True)

In [26]:
#Check the columns again
df.columns

Index(['LNG Flow', 'LNG Rundown Temperature', 'MR Operating Margin',
       'PR Operating Margin', 'MR Compressor Suction Pressure',
       'MCHE Warm bypass Temperature', 'MR return Temperature',
       'MR Expander inlet Temperature', 'LNG Shell Temp Top Bundle',
       'LMR JT Valve Valve Opening', 'MCHE Cold bundle dp',
       'Volumetric MR Flow to MR comp. discharge', 'HMR Flow ', 'HMR/MR Ratio',
       'LNG Flow Ex MCHE', 'MCHE Warm Bypass Temperature', 'MR IGV Position',
       'MR Vessel Level', 'MR Vessel Pressure', 'Percentage Nitrogen in MR',
       'Percentage Methane in MR', 'Percentage Ethane in MR',
       'Percentage Propane in MR', 'Percentage Butane in MR',
       'Nitrogen Make-up flow', 'Methane Make-up flow', 'Ethane Make-up flow',
       'Propane Make-up flow', 'Ambient Condition',
       'Scrub Column inlet Temperature',
       'Natural Gas temp at MCHE warm bundle',
       'Low Pressure Propane kettle level',
       'Medium Pressure Propane/Natural Gas kettle l

### -Drop all rows with null values

In [27]:
df.dropna(inplace=True)

In [28]:
df.shape

(32024, 40)

After removing all the null values, the number of records in our data frame has reduced to `32024` records

## Summary

In [29]:
# lets run a summary on the data 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32024 entries, 36328 to 70128
Data columns (total 40 columns):
 #   Column                                                     Non-Null Count  Dtype  
---  ------                                                     --------------  -----  
 0   LNG Flow                                                   32024 non-null  float64
 1   LNG Rundown Temperature                                    32024 non-null  float64
 2   MR Operating Margin                                        32024 non-null  float64
 3   PR Operating Margin                                        32024 non-null  float64
 4   MR Compressor Suction Pressure                             32024 non-null  float64
 5   MCHE Warm bypass Temperature                               32024 non-null  float64
 6   MR return Temperature                                      32024 non-null  float64
 7   MR Expander inlet Temperature                              32024 non-null  float64
 8   LN

## Export Data

The data looks neat, let's export the data and proceed to Exploratory data analysis

In [31]:
df.to_csv("LNG-DATA.csv")
print("Exported successfully")

Exported successfully
